Backend-tests.md

```diff
@@ -0,0 +1,596 @@
+
+  func_being_tested()
+  ```
+
+  There are also other swap functions you can use. They are defined and documented in [`core/tests/test_utils.py`](https://github.com/oppia/oppia/blob/develop/core/tests/test_utils.py).
+
+### Guidelines for writing good tests
+
+1. **Each test method should test only a single behaviour.** This helps both with naming the test and with ensuring that the test doesn't fail for unrelated code changes.
+
+   * When naming a test, start by writing a full sentence that clearly describes its behaviour. Try not to abbreviate if possible, but, if you need to do so in order to fit within the 80-character limit, make sure that the resulting test name is still meaningful and easy to understand.
+
+   * We recommend that test names follow the format:
+
+     ```text
+     test_{{action}}_with_{{with_condition}}_{{has_expected_outcome}}
+     ```
+
+     where `{{action}}`, `{{with_condition}}`, and `{{has_expected_outcome}}` are replaced with appropriate descriptions. Put the outcome at the end, so that it's easy to compare consecutive tests that have slightly different conditions and divergent outcomes. Here are some examples that follow this format:
+
+     * `test_get_by_auth_id_with_invalid_auth_method_name_is_none`
+     * `test_get_by_auth_id_for_unregistered_auth_id_is_empty_list`
+
+     These names are good because it's easy to see the differences between the tests: one tests an invalid auth method, and the other tests an unregistered auth. Correspondingly, these conditions lead to different outcomes (`name_is_none' vs. 'id_is_empty_list').
+
+2. Tests should use the following general structure:
+
+   * **Setup** - this is where you prepare any inputs/environment needed for the test.
+   * **Baseline verification** - check the values without performing any action. This step is only needed if your action is state-changing (i.e. if the same assert statement would lead to one result at the baseline and a different result at the endline). Check the same values here as you check at the endline.
+   * **Action** - perform the action or function call that leads to the expected change.
+   * **Endline verification** - check that the values from the baseline verification have changed accordingly.
+
+3. **Test the interface, not the implementation.** That is, treat the function as a black box and test its behavior. Don't test that the function uses a particular implementation. This will help you design a better API from an external user’s perspective. For example, consider the following function:
+
+   ```python
+   def absolute_value(number):
+       return abs(number)
+   ```
+
+   Forget for the moment that this function is pointless and think about how to test it. We should test the function by providing some values, some positive and some negative, and checking that their absolute values are returned. We should _not_ test it by mocking `abs()` and making sure it was called correctly.
```

Comment:
The rationale is stated in the first paragraph of the bullet:

> This will help you design a better API from an external user’s perspective.

By the "explanation ... delayed until the very end" I think you mean this part:

>  After changing the implementation, your tests should all still pass. In our example, our tests that provide various numbers would still pass. Our test that mocked `abs()` would not.

This isn't really the rationale. It's explaining how to apply the rule of thumb I introduce:

> You can check whether you're following this principle by imagining what would happen if you changed the function.

Reply:
Oh no -- to be clear, by "rationale" I mean the specific rationale for why "we should _not_ test it by mocking `abs()` and making sure it was called correctly." At the moment this is just asserted without explanation.

Reply:
done